synapse/lib/remcycle.py

```diff
@@ -0,0 +1,876 @@
+          service.  This string will be run through format() twice - once
+          during the construction of the Nyx object, and the second time
+          during the construction of the per-request url. As such, any
+          values set by the api configuration method noted below should
+          be encloused with double curly brackets.
+
+    The following configuration values are optional.
+
+        * api: This is a list of two objects, one listing required API values
+          and the second listing default API values.
+            - The first object should be a list of required values.  These
+              must be provided by the user when they call build_http_request.
+            - The second object should be a dictionary of optional values and
+              their defaults.  A user may provide alternative values when
+              calling build_http_request, but sensible defaults should be
+              provided here.
+        * http: A dictionary of key/value items which can provide per-api
+          specific arguements for the creation of HTTPRequest objects.
+        * ingests: A dictionary of Synapse ingest definitions which will be
+          used to create Ingest objects.  During registration of a Nyx object
+          with Hypnos, these will be registered into the Hypnos cortex.
+          Multiple named ingests may be made available for a single API.
+        * vars: This is a list of 2 value items which are stamped into the url
+          value during the construction of the Nyx object.
+
+    See a complete example below:
+
+    ::
+
+        {
+          "api": [
+            [
+              "someplace"
+            ],
+            {
+              "domore": 0
+            }
+          ],
+          "doc": "api example",
+          "http": {
+            "headers": {
+              "token-goodness": "sekrit token"
+            }
+          },
+          "ingests": {
+            "ingest_definition": {
+              "ingest": {
+                "forms": [
+                  [
+                    "inet:ipv4",
+                    {
+                      "var": "ip"
+                    }
+                  ]
+                ],
+                "vars": [
+                  [
+                    "ip",
+                    {
+                      "path": "ip"
+                    }
+                  ]
+                ]
+              }
+            }
+          },
+          "url": "http://vertex.link/api/v4/geoloc/{{someplace}}/info?domore={{domore}}&apikey={APIKEY}",
+          "vars": [
+            ["APIKEY", "8675309"]
+          ]
+        }
+
+    ::
+
+    This example should be interpreted as the following:
+
+        1. The APIKEY value in the 'vars' will be set in the URL, resulting in
+           the following default url:
+
+        ::
+
+            "http://vertex.link/api/v4/geoloc/{someplace}/info?domore={domore}&apikey=8675309"
```

Comment:
There are a few different ways to skin this cat.  Currently, the url is passed through .format() twice - once when the Nyx object is created, and then each time a HTTPRequest object is created.  This requires any values which we want to set the second time to be defined with double curlys; and anything we set the first time to have single curlys.  An alternative is to do a single pass of format() during the HTTPRequest object creation, which means every user provided value and fixed value would have to have single curlys.

.format() as chosen over string.Template(), since its considerably faster without the potential headache caused with %s formatting.  See http://tomatohater.com/2013/08/30/python-performance-string-formatting/ for some performance metrics.